Structured kernels generate Meta registrations

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48687 Delete NativeFunctions.h include from Functions.h
* #48683 Move var and std overloads to Functions.cpp and remove native:: reference
* #48659 Refactor TensorIterator to do allocations via MetaBase::set_output
* #48195 Move argument grouping into FunctionSchema
* #48182 Refactor argument fields in FunctionSchema to Arguments
* **#48116 Structured kernels generate Meta registrations**

If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25056640](https://our.internmc.facebook.com/intern/diff/D25056640)

ezyang committed:
Structured kernels generate Meta registrations

If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Structured kernels generate Meta registrations"


If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -14,6 +14,11 @@ struct TensorMeta {
+}
+
 inline Tensor tensor_from_meta(const TensorMeta& meta) {
   // TODO: eliminate indirection
```
What's the downside of this indirection? I think the wrapper function name makes it a little clearer what the goal is in the codegen.

ezyang replied:
It's actually a different indirection; `at::empty_meta` is making a hop in the dispatcher but it's unnecessary, we could hardcode `at::native::empty_meta` here. I haven't done it yet because I want to keep implementation simple for now.

bdhirsh replied:
ahhh yep

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1672,6 +1672,7 @@
     CPU: resize_
     CUDA: resize_cuda_
     QuantizedCPU: quantized_resize_cpu_
+    Meta: resize_meta_
```
This means that we can't make resize_ `structured`, right? You have a check in `gen_structured` to assert that people don't explicitly include their own structured implementations

bdhirsh replied:
Possible alternative: Just call the meta function (`resize_meta_`) directly rather than through the dispatcher in our codegen. I think that just means that we'll need to call `meta_tensor_from_meta` to convert it to a tensor directly in the codegen as well.

bdhirsh replied:
aside: I'm not clear on the use case for `out`-variant `meta` functions. If your goal is to find the meta-info of the output tensor, you can just examine `out` argument directly, right?

I could imagine the goal being to take an existing model, stick the `Meta` dispatch key on all the input tensors, and run anything without having to make changes to the model. That's probably it.

ezyang replied:
> This means that we can't make resize_ structured, right? 

This is correct. In fact, we can't make `resize_` structured anyway, because there is no out variant nor purely functional variant (nor would they make sense).

> Possible alternative: Just call the meta function (resize_meta_) directly rather than through the dispatcher in our codegen.

We should do this anyway.

> I could imagine the goal being to take an existing model, stick the Meta dispatch key on all the input tensors, and run everything without having to make changes to the model. That's probably it.

Yep!

bdhirsh commented on test/test_torch.py:
```diff
@@ -4483,6 +4483,20 @@ def test_empty_meta(self):
+            # (not sure why passing None here doesn't work? How strange...)
+            z = torch.empty_meta(0)
+            torch._C._nn.upsample_nearest1d(x, (4 * 10 ** 10,), 2, out=z)
+            self.assertEqual(z.size(), (2 * 10 ** 10, 3, 4 * 10 ** 10))
```
maybe worth asserting that we didn't actually perform the computation? Also totally understand that the meta testing is subject to change.

ezyang replied:
I'm not... sure exactly how to test that haha. I'm sort of trying to implicitly test this by using ludicrously large sizes here; if you actually allocated these you'd probably OOM.

bdhirsh replied:
I was just thinking running the meta op, running the actual op, then asserting that their `out` results aren't equal (I'm not actually sure what the contents of the `empta_meta` are- if it's uninitialized memory we could just zero it out first).

bhosmer replied:
A meta tensor should give a predictable error if you try to access its data, right? Can we just test by expecting that error here?

In master `torch.empty_meta(10)[0]` gets you a RuntimeError:
```
RuntimeError: Could not run 'aten::as_strided' with arguments from the 'Meta' backend.
```

Expecting a RuntimeError on data access (not on `as_strided`, just in general) seems like a reasonable proxy for "didn't compute"

ezyang committed:
Update on "Structured kernels generate Meta registrations"


If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25056640](https://our.internmc.facebook.com/intern/diff/D25056640)

[ghstack-poisoned]

ezyang commented:
The ASAN failure on this PR is legit

ezyang committed:
Update on "Structured kernels generate Meta registrations"


If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25056640](https://our.internmc.facebook.com/intern/diff/D25056640)

[ghstack-poisoned]

ezyang committed:
Update on "Structured kernels generate Meta registrations"


If you port kernels to be structured, you get Meta kernels automatically
generated for you.  This is one payoff of structured kernels.

Code generation was mercifully really simple, although at risk of
"swiss cheese" syndrome: there's two new conditionals in the codegen
to tweak behavior when generating for meta keys.  It's not too bad
right now but there's a risk of things getting out of hand.  One
way to rationalize the logic here would be to transmit "TensorMeta-ness"
inside the TensorOptions (so tensor_from_meta can deal with it); then
the "Meta" kernel magic would literally just be generating empty
out_impls to call after all the scaffolding is done.  But I didn't
do this because it seemed like it would be more annoying short term.

Also had to teach resize_ to work on meta tensors, since we use them
to implement the out kernels.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25056640](https://our.internmc.facebook.com/intern/diff/D25056640)

[ghstack-poisoned]
